pkg/reconciler/awssnssource/subscription.go

```diff
@@ -127,39 +143,48 @@ func (r *Reconciler) ensureUnsubscribed(ctx context.Context) error {
 	case isNotFound(err):
 		// the finalizer is unlikely to recover from a missing Secret,
 		// so we simply record a warning event and return
-		event.Warn(ctx, ReasonFailedUnsubscribe, "Secret missing while finalizing subscription %q. Ignoring: %s",
-			*subsARN, err)
+		event.Warn(ctx, ReasonFailedUnsubscribe,
+			"Secret missing while finalizing event source. Ignoring: %s", err)
 		return nil
 	case err != nil:
 		return fmt.Errorf("%w", reconciler.NewEvent(corev1.EventTypeWarning, ReasonFailedUnsubscribe,
 			"Error creating SNS client: %s", err))
 	}
 
-	resp, err := snsClient.UnsubscribeWithContext(ctx, &sns.UnsubscribeInput{
-		SubscriptionArn: subsARN,
-	})
+	topicARN := spec.ARN.String()
 
+	subsARN, err := findSubscription(ctx, snsClient, topicARN, url.String())
 	switch {
 	case isNotFound(err):
 		return reconciler.NewEvent(corev1.EventTypeNormal, ReasonUnsubscribed,
-			"Subscription %q already absent, skipping finalization", *subsARN)
+			"Subscription already absent, skipping finalization")
 	case isDenied(err):
 		// it is unlikely that we recover from validation errors in the
 		// finalizer, so we simply record a warning event and return
-		event.Warn(ctx, ReasonFailedUnsubscribe, "Authorization error finalizing subscription %q. Ignoring: %s",
-			*subsARN, toErrMsg(err))
+		event.Warn(ctx, ReasonFailedUnsubscribe,
+			"Authorization error finding subscription. Ignoring: %s", toErrMsg(err))
 		return nil
 	case err != nil:
 		// wrap any other error to fail the finalization
-		event := reconciler.NewEvent(corev1.EventTypeWarning, ReasonFailedUnsubscribe,
-			"Error finalizing event source %q: %s", *subsARN, toErrMsg(err))
-		return fmt.Errorf("%w", event)
+		return fmt.Errorf("%w", reconciler.NewEvent(corev1.EventTypeWarning, ReasonFailedUnsubscribe,
+			"Error finding subscription: %s", toErrMsg(err)))
 	}
 
-	logging.FromContext(ctx).Debug("Unsubscribe responded with: ", resp)
+	err = unsubscribe(ctx, snsClient, subsARN)
+	switch {
+	case isDenied(err):
+		// it is unlikely that we recover from validation errors in the
+		// finalizer, so we simply record a warning event and return
+		event.Warn(ctx, ReasonFailedUnsubscribe, "Authorization error unsubscribing subscription %q "+
+			"from topic %q. Ignoring: %s", subsARN, topicARN, toErrMsg(err))
+		return nil
+	case err != nil:
+		// wrap any other error to fail the finalization
+		return fmt.Errorf("%w", unsubscribeErrorEvent(subsARN, topicARN, err))
+	}
```

Comment:
There are a bunch of those error switches scattered across the `ensureSubscribed()` and `ensureUnsubscribed()` methods,  and they are pretty verbose, but I still like that better than using generic error handling helpers because it allows us to return specific errors based on the situation (e.g. we handle things slightly differently depending on whether the error occurs during a lookup or a Subscribe call).